bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,1139 @@
+                if (Objects.nonNull(responseSuccess)) {
+                    String newApplicationKey = (String) responseSuccess.get("username");
+                    if (Objects.nonNull(newApplicationKey)) {
+                        return newApplicationKey;
+                    }
+                }
+            }
+        } catch (JsonParseException e) {
+            logger.debug("registerApplicationKey() error parsing JSON response:{}", json);
+        }
+        throw new HttpUnauthorizedException("Application key registration failed");
+    }
+
+    public void setExternalRestartScheduled() {
+        externalRestartScheduled = true;
+        internalRestartScheduled = false;
+        internalRestartTask = cancelTask(internalRestartTask, false);
+        close2();
+    }
+
+    /**
+     * Sleep the caller during any period when the connection is restarting.
+     *
+     * @throws ApiException if anything failed.
+     * @throws InterruptedException
+     */
+    private void sleepDuringRestart() throws ApiException, InterruptedException {
+        Future<?> restartTask = this.internalRestartTask;
+        if (Objects.nonNull(restartTask)) {
+            try {
+                restartTask.get(RESTART_AFTER_SECONDS * 2, TimeUnit.SECONDS);
+            } catch (ExecutionException | TimeoutException e) {
+                throw new ApiException("sleepDuringRestart() error", e);
+            }
+        }
+        internalRestartScheduled = false;
+    }
+
+    /**
+     * Test the Hue Bridge connection state by attempting to connect and trying to execute a basic command that requires
+     * authentication.
+     *
+     * @throws HttpUnauthorizedException if it was possible to connect but not to authenticate.
+     * @throws ApiException if it was not possible to connect.
+     * @throws InterruptedException
+     */
+    public void testConnectionState() throws HttpUnauthorizedException, ApiException, InterruptedException {
+        logger.debug("testConnectionState()");
+        try {
+            openPassive();
+            getResourcesImpl(BRIDGE);
+        } catch (ApiException e) {
+            close2();
+            throw e;
+        }
+    }
+
+    /**
+     * Hue Bridges get confused if they receive too many HTTP requests in a short period of time (e.g. on start up), or
+     * if too many HTTP sessions are opened at the same time. So this method throttles the requests to a maximum of one
+     * per REQUEST_INTERVAL_MILLISECS, and ensures that no more than MAX_CONCURRENT_SESSIONS sessions are started.
+     *
+     * @throws InterruptedException
+     */
+    private synchronized void throttle() throws InterruptedException {
+        try {
+            streamMutex.acquire();
+            long delay = Duration.between(Instant.now(), lastRequestTime).toMillis() + REQUEST_INTERVAL_MILLISECS;
+            if (delay > 0) {
+                Thread.sleep(delay);
+            }
+        } catch (ArithmeticException e) {
```

Comment:
There is no need to catch this `RuntimeException`, except if you could actually exceed the capacity of `Duration` with the calculation above. But then we would have much worse problems. This would hide a bug while bypassing the throttler.

Reply:
Ok. Done.

Reply:
Ok. Fixed again.

Reply:
> Wouldn't it be more elegant to avoid the exception though?

You beat me, was about to write same comment.

Reply:
> make lastRequestTime an Optional<Instant> 

And fixed. Again. Again.

Reply:
Yet another improvement suggestion (shorter and makes timestamp assignment atomic):

```java
private synchronized void throttle() throws InterruptedException {
    streamMutex.acquire();
    final Instant now = Instant.now();
    lastRequestTime.ifPresent(t -> {
        long delay = Duration.between(now, t).toMillis() + REQUEST_INTERVAL_MILLISECS;
        Thread.sleep(delay);
    });
    lastRequestTime = now;
}
```

Reply:
Cool code idea :)
But it doesn't compile because 'Thread.sleep()' can throw 'InterruptedException' but the exception cannot be pushed up the stack from inside the 'ifPresent()' call -- at least not without adding more code than you would have saved :(